fiona/model.py

```diff
@@ -17,31 +53,268 @@ class Object(MutableMapping):
     about future deprecation of features.
     """
 
-    def __init__(self, **data):
-        self._data = {}
-        self._data.update(data)
+    _delegated_properties = []
+
+    def __init__(self, **kwds):
+        self._data = OrderedDict(**kwds)
+
+    def _props(self):
+        return {
+            k: getattr(self._delegate, k)
+            for k in self._delegated_properties
+            if k is not None
+        }
 
     def __getitem__(self, item):
-        return self._data[item]
+        props = self._props()
+        props.update(**self._data)
+        return props[item]
 
     def __iter__(self):
-        return iter(self._data)
+        props = self._props()
+        return itertools.chain(iter(props), iter(self._data))
 
     def __len__(self):
-        return len(self._data)
+        props = self._props()
+        return len(props) + len(self._data)
 
     def __setitem__(self, key, value):
         warn(
             "instances of this class -- CRS, geometry, and feature objects -- will become immutable in fiona version 2.0",
             FionaDeprecationWarning,
             stacklevel=2,
         )
-        self._data[key] = value
+        if key in self._delegated_properties:
+            setattr(self._delegate, key, value)
+        else:
+            self._data[key] = value
 
     def __delitem__(self, key):
         warn(
             "instances of this class -- CRS, geometry, and feature objects -- will become immutable in fiona version 2.0",
             FionaDeprecationWarning,
             stacklevel=2,
         )
-        del self._data[key]
+        if key in self._delegated_properties:
+            setattr(self._delegate, key, None)
+        else:
+            del self._data[key]
+
+
+class _Geometry(object):
+    def __init__(self, coordinates=None, type=None):
+        self.coordinates = coordinates
+        self.type = type
+
+
+class Geometry(Object):
+    """A GeoJSON-like geometry
+
+    Notes
+    -----
+    Delegates coordinates and type properties to an instance of
+    _Geometry, which will become an extension class in Fiona 2.0.
+
+    """
+
+    _delegated_properties = ["coordinates", "type"]
+
+    def __init__(self, coordinates=None, type=None, **data):
+        self._delegate = _Geometry(coordinates=coordinates, type=type)
+        super(Geometry, self).__init__(**data)
+
+    @classmethod
+    def from_dict(cls, **kwds):
```

Comment:
This is actually not "from a dict", in the sense that with the current implementation you cannot pass a dictionary:

```
geom = {'type': 'Point', 'coordinates': [...]}
Geometry.from_dict(geom)
```

will not work. `Geometry.from_dict(**geom)` of course does, but the name might imply otherwise. 

(just something I noticed while going through the code, not that you need to change anything ;) I also don't know how public those classes will be?)

Reply:
It would not be hard to take dict or **dict. Thanks for the feedback and testing, @jorisvandenbossche !